### PR TITLE
fix(ReactContext): Dispose path for ReactContext created deadlock

### LINKED ISSUE #702: Code Push
I have not been able to get codepush working for windows. Android, iOS no problem

Microsoft/code-push#306

### PR DESCRIPTION
**Pull request by rozele**
Basic overview of the deadlock:
1) CSS layout operation in progress
2) Text node measurement started on native module thread
3) ReactContext disposed on dispatcher thread
4) ReactContext disposal blocks on completion of native module thread
5) Native module thread blocks on dispatcher operation for text measurement
6) We have deadlock

The best way to get around this is to not block from the native module thread, but this is a limitation we haven't figured out how to work around yet.  In the meantime, it actually makes sense to make the teardown process for the React context asynchronous, which is what this changeset does.

Fixes #702

### CODE REVIEW

**Comment by rozele** [comment]
cc @mpodwysocki @andrewwebber @matthargett 

This is not a trivial change so I'd love to get some feedback.

**Comment by matthargett** [comment]
will review more deeply once I drop off my daughter at day care, but my first question is: can the hang be reproduced in a unit/integration test?

**Comment by andrewwebber** [comment]
I can create a same app when required. Unfortunately I am at the Oktoberfest for the whole weekend so I will probably be able to build a sample on Monday

File: ReactWindows/ReactNative/ReactInstanceManager.cs [hunk 1]
```diff
@@ -290,7 +291,7 @@ public void OnDestroy()
             var currentReactContext = _currentReactContext;
             if (currentReactContext != null)
             {
-                currentReactContext.Dispose();
+                await currentReactContext.DisposeAsync();
```
Comment (rozele) [inline ReactWindows/ReactNative/ReactInstanceManager.cs]: ConfigureAwait(false) should be okay here.

File: ReactWindows/ReactNative/Bridge/ReactInstance.cs [hunk 2]
```diff
@@ -179,13 +179,13 @@ public void Dispose()
             IsDisposed = true;
             _registry.NotifyReactInstanceDispose();
 
-            QueueConfiguration.JavaScriptQueueThread.CallOnQueue(() =>
+            await QueueConfiguration.JavaScriptQueueThread.CallOnQueue(() =>
```
Comment (rozele) [inline ReactWindows/ReactNative/Bridge/ReactInstance.cs]: Can these be done concurrently?

File: ReactWindows/ReactNative/Bridge/ReactInstance.cs [hunk 3]
```diff
@@ -215,7 +215,7 @@ private string BuildModulesConfig()
         private void HandleException(Exception ex)
         {
             _nativeModuleCallExceptionHandler(ex);
-            QueueConfiguration.DispatcherQueueThread.RunOnQueue(Dispose);
+            QueueConfiguration.DispatcherQueueThread.RunOnQueue(async () => await DisposeAsync());
```
Comment (rozele) [inline ReactWindows/ReactNative/Bridge/ReactInstance.cs]: ConfigureAwait

File: ReactWindows/ReactNative.Tests/ReactInstanceManagerTests.cs [hunk 4]
```diff
@@ -174,14 +174,55 @@ public void ReactInstanceManager_Builder_SetterChecks()
...
+            var tcs = new TaskCompletionSource<bool>();
+            await DispatcherHelpers.RunOnDispatcherAsync(async () =>
+            {
+                e.Set();
+                await manager.DisposeAsync();
+                await Task.Run(() => tcs.SetResult(true));
+            });
+
+            var completedTask = await Task.WhenAny(
+                Task.Delay(50000),
```
Comment (rozele) [inline ReactWindows/ReactNative.Tests/ReactInstanceManagerTests.cs]: >50000 [](start = 27, length = 5)

Shorten.

File: ReactWindows/ReactNative/Bridge/ReactInstance.cs (hunk 2 above)
Comment (rozele) [inline ReactWindows/ReactNative/Bridge/ReactInstance.cs]: Silly question, would introduce a race where we dispose the JS queue thread before we can use it to dispose the bridge.

---
In reply to: [80536408](https://github.com/ReactWindows/react-native-windows/pull/712#discussion_r80536408) [](ancestors = 80536408)

**Pull request merged**